File: java/src/main/java/recipes/OhlcvToCsv.java
// Download paginated OHLCV candles and write them to a CSV file.
package recipes;

import io.github.ccxt.exchanges.Binance;
import io.github.ccxt.types.OHLCV;

import java.io.PrintWriter;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

public class OhlcvToCsv {

    public static void main(String[] args) throws Exception {
        String symbol = args.length > 0 ? args[0] : "BTC/USDT";
        String timeframe = args.length > 1 ? args[1] : "1h";
        int days = args.length > 2 ? Integer.parseInt(args[2]) : 7;

        Binance exchange = new Binance();
        long cursor = System.currentTimeMillis() - (long) days * 86400 * 1000;
        List<OHLCV> candles = new ArrayList<>();
        while (true) {
            List<OHLCV> batch = exchange.fetchOHLCV(symbol, timeframe, cursor, 1000L);
            if (batch.isEmpty()) {
                break;
            }
            candles.addAll(batch);
            long lastTimestamp = batch.get(batch.size() - 1).timestamp;
            if (batch.size() < 1000 || lastTimestamp <= cursor) {
                break; // reached the newest candle
            }
            cursor = lastTimestamp + 1;
        }

        String filename = symbol.replace("/", "-").replace(":", "-") + "-" + timeframe + ".csv";
        try (PrintWriter writer = new PrintWriter(filename)) {
            writer.println("timestamp,datetime,open,high,low,close,volume");
            for (OHLCV c : candles) {
                writer.printf("%d,%s,%s,%s,%s,%s,%s%n",
                    c.timestamp, Instant.ofEpochMilli(c.timestamp), c.open, c.high, c.low, c.close, c.volume);
            }
        }
        System.out.println(candles.size() + " candles -> " + filename);
    }
}

File: csharp/recipes/OhlcvToCsv.cs
// Download paginated OHLCV candles and write them to a CSV file.
using System.Text;
using ccxt;

namespace Recipes;

public static class OhlcvToCsv
{
    public static async Task Run(string[] args)
    {
        var symbol = args.Length > 0 ? args[0] : "BTC/USDT";
        var timeframe = args.Length > 1 ? args[1] : "1h";
        var days = args.Length > 2 ? int.Parse(args[2]) : 7;

        var exchange = new Binance();
        var cursor = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - (long)days * 86400 * 1000;
        var candles = new List<OHLCV>();
        while (true)
        {
            var batch = await exchange.FetchOHLCV(symbol, timeframe, cursor, 1000);
            if (batch.Count == 0)
            {
                break;
            }
            candles.AddRange(batch);
            var lastTimestamp = (long)batch[^1].timestamp!;
            if (batch.Count < 1000 || lastTimestamp <= cursor)
            {
                break; // reached the newest candle
            }
            cursor = lastTimestamp + 1;
        }

        var filename = $"{symbol.Replace("/", "-").Replace(":", "-")}-{timeframe}.csv";
        var sb = new StringBuilder("timestamp,datetime,open,high,low,close,volume\n");
        foreach (var c in candles)
        {
            var datetime = DateTimeOffset.FromUnixTimeMilliseconds((long)c.timestamp!).UtcDateTime.ToString("o");
            sb.AppendLine($"{c.timestamp},{datetime},{c.open},{c.high},{c.low},{c.close},{c.volume}");
        }
        File.WriteAllText(filename, sb.ToString());
        Console.WriteLine($"{candles.Count} candles -> {filename}");
    }
}
